lib/money/money/allocation.rb

```diff
@@ -0,0 +1,54 @@
+# encoding: utf-8
+
+class Money
+  class Allocation
+    # Splits a given amount in parts without loosing pennies.
+    # The left-over pennies will be distributed round-robin amongst the parties. This means that
+    # parties listed first will likely receive more pennies than ones that are listed later.
+    #
+    # The results should always add up to the original amount.
+    #
+    # The parts can be specified as:
+    #   Numeric — performs the split between a given number of parties evenely
+    #   Array<Numeric> — allocates the amounts proportionally to the given array
+    #
+    def initialize(amount, parts, whole_amounts = true)
+      @amount = amount
+      @parts = parse_parts(parts)
+      @whole_amounts = whole_amounts
+
+      raise ArgumentError, 'need at least one party' if @parts.empty?
+    end
+
+    def generate
+      result = []
+      remaining_amount = amount
+      remaining_parts = parts.reverse
+
+      while !remaining_parts.empty? do
+        parts_sum = remaining_parts.inject(0, :+)
+        part = remaining_parts.shift
+
+        current_split = remaining_amount * part / parts_sum
+        current_split = current_split.truncate if whole_amounts
+
+        result << current_split
+        remaining_amount -= current_split
+      end
+
+      result.reverse
+    end
```

Comment:
I think the strategy is brilliant! 🏆

I tried to come up with a "less imperative" version, but could only golf it as far as this (which I don't really like because of the indexing mess):

```ruby
    def generate
      remaining_amount = amount

      parts.reverse.map.with_index do |part, i|
        current_split = remaining_amount * part / parts[0..(-i-1)].inject(0, :+)
        current_split = current_split.truncate if whole_amounts

        remaining_amount -= current_split

        current_split
      end.reverse
    end
```

Reply:
Should `.ceil` allow to avoid double reverse?

Reply:
Not bad thinking! However, the `reverse` is there to leave the "extra pennies" (that are left "in the end" of the `map`) to the first parts.

For instance, if we are to divide `amount = 10` into `parts = [1,1,2]` (spec in line 83); the `map` "loops" would have this before subtracting from `remaining_amount`:
```
[part, i, remaining_amount.to_s, current_split.to_s]
[2   , 0, "10"                 , "5"]
[1   , 1, "5"                  , "2"]
[1   , 2, "3"                  , "3"]
```

Or, slightly more intricate, if we divide `amount = BigDecimal(10)` into `parts = [1,1,1]` with `whole_amounts = false` (spec in line 105):
```
[part, i, remaining_amount.to_s      , current_split.to_s]
[1   , 0, "0.1E3"                    , "0.33333333333333333333E2"]
[1   , 1, "0.66666666666666666667E2" , "0.333333333333333333335E2"]
[1   , 2, "0.333333333333333333335E2", "0.333333333333333333335E2"]
```

By removing both `reverse`s and replacing `truncate` with `ceil` (and replacing `parts[0..(-i-1)]` with `parts[i..-1]` and `#to_f`-ing the division because integer division by default truncates) we get
```
[1, 0, "10", "2"]
[1, 1, "8", "2"]
[2, 2, "6", "6"]
```
which is wrong (according to the spec that the extra pennies to go to the first parts) and
```
[1, 0, "0.1E3"                    , "0.33333333333333333333E2"]
[1, 1, "0.66666666666666666667E2" , "0.333333333333333333335E2"]
[1, 2, "0.333333333333333333335E2", "0.333333333333333333335E2"]
```
which is also wrong because it's not afterwards reversed; thus, the last two parts receives the extra.

In other words, if the spec was to send the extra pennies to the last, you'd be right.

Reply:
Please take a look: https://gist.github.com/printercu/f9dbde27172ca4a57533d1dcdd95f2a3

It works the same way: greater values appear in the begining. But it gives different results, and it's not clear for me which is better. For example, split 10 into 1, 1, 2: `[3, 2, 5]` or `[3, 3, 4]`.

Reply:
> (...) and it's not clear for me which is better. (...)

Me neither. I expect it's possible to find cases where the cumulative relative error will favour any of the two. If that's correct, then it's just an arbitrary choice between the two, and in that case, I agree that the current spec should be favoured.

I guess an alternative strategy to the double reverse could be "Work your way from below and mirror around zero", say:

* "From below": `remaining_amount = -amount` and `+=` (rather than `-=`; for negative numbers, `ceil` is the same as `truncate` so choose either)
* "Mirror": `.reverse.map{|x| x+amount}`

I suspect the performance is quite alike.

Speaking about performance, as long as the complexity is in terms of `parts.length` I wouldn't care much. If users have "too long `parts`" then they may optimise and contribute 😉

I wonder if anyone is actually using this `split` method? I notice that https://github.com/carlospalol/money/ does not have such a thing 🙂

Reply:
No idea if anyone is actually using this method :) I can definitely see how it can be extremely useful in certain situations, but that's definitely not your typical scenario.

Reply:
`.lazy` before first `.reverse` may help avoiding actual reverses. But I don't know if it's available in all rubies supported by gem.

Reply:
I've modified it so it now doesn't have any `.reverse` calls